[UNI-299] Implementation of paramFinder and paramFinder runner

ywcui1990 opened the pull request:
@marionleborgne @vitaly-krugl 

A first version of paramFinder class (for future nupic use) and the paramFinder runner (for unicorn)

I am still working on the tests. But any early feedback on this implementation is appreciated!

marionleborgne commented:
:+1: I ran `param_finder_runner.py` on the hotgym data and it gave me a windows size of 1.5 hour for aggregation with the "mean" operation. 
Adding a unit test to validate that the param_finder returns the right params will be good indeed. Let me know when you've added that.

marionleborgne commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -0,0 +1,163 @@
+               timestampIndex=0,
+               valueIndex=1):
+    """
+    :param str modelId: model identifier
```
This is the old doc string from `model_runner.py` :smile: Please update the param names and descriptions.

marionleborgne commented:
@ywcui1990 Below is the JSON returned by the param finder. I have a couple of questions:
1. Why is `predictionSteps` set to 1? We only want to do anomaly detection
2. Why is `predictedField` set to `c1`? I ran the param finder against a CSV where the metric name is `attendeeCount` not `c1`. Are you sure the field name gets updated correctly? I see a couple of `c1` in the JSON returned, instead of the actual metric name. Also, @vitaly-krugl how do we handle the case where a CSV file without headers is passed (in that case, what is the name of the predicted field? is it `c1`?)


```
         {
           "aggInfo":{  
               "windowSize":5270,
               "func":"mean"
            },
            "modelInfo":{  
               "inferenceArgs":{  
                  "predictionSteps":[  
                     1
                  ],
                  "predictedField":"c1",
                  "inputPredictedField":"auto"
               },
               "modelConfig":{  
                  "aggregationInfo":{  
                     "seconds":0,
                     "fields":[  

                     ],
                     "months":0,
                     "days":0,
                     "years":0,
                     "hours":0,
                     "microseconds":0,
                     "weeks":0,
                     "minutes":0,
                     "milliseconds":0
                  },
                  "model":"CLA",
                  "version":1,
                  "predictAheadTime":null,
                  "modelParams":{  
                     "sensorParams":{  
                        "verbosity":0,
                        "encoders":{  
                           "c0_dayOfWeek":{  
                              "dayOfWeek":[  
                                 21,
                                 3
                              ],
                              "type":"DateEncoder",
                              "fieldname":"c0",
                              "name":"c0"
                           },
                           "c0_timeOfDay":null,
                           "c1":{  
                              "name":"c1",
                              "resolution":3.2857343580170375,
                              "seed":42,
                              "fieldname":"c1",
                              "type":"RandomDistributedScalarEncoder"
                           },
                           "c0_weekend":null
                        },
                        "sensorAutoReset":null
                     },
                     "clEnable":false,
                     "spParams":{  
                        "columnCount":2048,
                        "spVerbosity":0,
                        "maxBoost":1.0,
                        "spatialImp":"cpp",
                        "inputWidth":0,
                        "synPermInactiveDec":0.0005,
                        "synPermConnected":0.1,
                        "synPermActiveInc":0.0015,
                        "seed":1956,
                        "numActiveColumnsPerInhArea":40,
                        "globalInhibition":1,
                        "potentialPct":0.8
                     },
                     "trainSPNetOnlyIfRequested":false,
                     "clParams":{  
                        "alpha":0.035828933612158,
                        "clVerbosity":0,
                        "steps":"1",
                        "regionName":"CLAClassifierRegion"
                     },
                     "tpParams":{  
                        "columnCount":2048,
                        "activationThreshold":13,
                        "pamLength":3,
                        "cellsPerColumn":32,
                        "permanenceInc":0.1,
                        "minThreshold":10,
                        "verbosity":0,
                        "maxSynapsesPerSegment":32,
                        "outputType":"normal",
                        "globalDecay":0.0,
                        "initialPerm":0.21,
                        "permanenceDec":0.1,
                        "seed":1960,
                        "maxAge":0,
                        "newSynapseCount":20,
                        "maxSegmentsPerCell":128,
                        "temporalImp":"cpp",
                        "inputWidth":2048
                     },
                     "anomalyParams":{  
                        "anomalyCacheRecords":null,
                        "autoDetectThreshold":null,
                        "autoDetectWaitRecords":5030
                     },
                     "spEnable":true,
                     "inferenceType":"TemporalAnomaly",
                     "tpEnable":true
                  }
               }
            }
         }
```

marionleborgne commented:
@vitaly-krugl based on our last discussion about the timestamp format, shouldn't we have one additional command line arg that allows us to pass the timestamp format string to the `param_finder_runner`?

vitaly-krugl commented:
@marionleborgne, regarding
> @vitaly-krugl could you remind me how do we handle the case where a CSV file without headers is passed (in that case, what is the name of the predicted field if there is no header row? is it `c1`?)

Actually, "c0" and "c1" (or "timestamp" and "value") are better than arbitrary column names from the csv file, which might contain spaces and other unsavory characters. What we name things on the backend side has no bearing on the computation, and removing user-related nuances for the list of problems to deal with is a good thing.

vitaly-krugl commented:
@marionleborgne, regarding
> @vitaly-krugl based on our last discussion about the timestamp format, shouldn't we have one additional command line arg that allows us to pass the timestamp format string to the `param_finder_runner`?

Yes, I think so. It's called `datetimeFormat` and is described in the "ParamFinder Command-line Args" section of [Unicorn Front End/Back End Interface (phase 2)](https://docs.google.com/document/d/1XZYGFCIoK7NarAl-DzCqluaTHac1PYFZQaEiu0r2rhs)

vitaly-krugl commented:
@marionleborgne, good catch regarding `timestampFieldName` and `valueFieldName`. Those need to be implemented per the "ParamFinder Command-line Args" section of [Unicorn Front End/Back End Interface (phase 2)](https://docs.google.com/document/d/1XZYGFCIoK7NarAl-DzCqluaTHac1PYFZQaEiu0r2rhs)

vitaly-krugl commented:
@marionleborgne and @ywcui1990, regarding Marion's question
>  Why is predictionSteps set to 1? We only want to do anomaly detection

see https://github.com/numenta/nupic/blob/master/src/nupic/frameworks/opf/common_models/anomaly_params_random_encoder/best_single_metric_anomaly_params.json, which is a template for the configuration parameters returned by `nupic.frameworks.opf.common_models.cluster_params.getScalarMetricWithTimeOfDayAnomalyParams()`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+from nupic.frameworks.opf.common_models.cluster_params import (
+  getScalarMetricWithTimeOfDayAnomalyParams)
+
+_modeFromNameDict = {
```
If it's intended to be a constant, then make it all-caps, separated by underscores: `_NAME_TO_MODE_MAP`.

If it's intended to be a module-level global variable, then prefix it with `g`: `g_nameToModelMap`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+Implements param finder, which automatically
+1. identify appropriate data aggregation window
+2. suggest whether to use TimeOfDay encoder
+3. suggest data aggregation type (SUM or AVG)
```
"sum" or "mean"

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+1. identify appropriate data aggregation window
+2. suggest whether to use TimeOfDay encoder
+3. suggest data aggregation type (SUM or AVG)
+"""
```
Thank's for putting a description in the docstring. This is helpful.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+import csv
+import dateutil.parser
+import numpy
+from nupic.frameworks.opf.common_models.cluster_params import (
```
Need whitespace between groups of imports. 0ne line to separate builtin from external imports in this case

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+
+def _modeFromName(mode):
```
This seems like unnecessary complexity. Instead of `_modeFromNameDict`, just define model-level constants, and pass those directly to `_convolve()` without the need to map. E.g.,

```
_CORRELATION_MODE_VALID = 0
_CORRELATION_MODE_SAME = 1
_CORRELATION_MODE_FULL = 2

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+
+def _convolve(a, v, mode='full'):
```
Short names like `a` and `v` are frowned upon in python.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+  Parameters
+  ----------
+  :param a : (N,) array_like
```
Should be no space between variable name and `:`

e.g., `:param a:`

vitaly-krugl commented:
Need to run the two scripts through pylint using pylintrc from `numenta-apps/pylintrc` for numenta-apps code and `nupic/pylintrc` for nupic code and fix findings.

vitaly-krugl commented:
Need unit tests. See https://github.com/numenta/numenta-apps/tree/master/unicorn/py/unicorn_backend/tests

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+  a, v = numpy.array(a, ndmin=1), numpy.array(v, ndmin=1)
+  if len(v) > len(a):
+    a, v = v, a
+  if len(a) == 0:
```
It's probably easier to understand, if you place the `==` checks before the `>` block, even though the result is the same in this function. The idea is that you want to perform arg validation before the core logic of the function.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+  .. [1] Wikipedia, "Convolution", http://en.wikipedia.org/wiki/Convolution.
+
+  """
+  a, v = numpy.array(a, ndmin=1), numpy.array(v, ndmin=1)
```
The code is more readable if you use individual statements on individual lines; there is no need to bundle them like this, making it more difficult to understand. The following is more readable:

```
a = numpy.array(a, ndmin=1)
v = numpy.array(v, ndmin=1)
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+  """
+  a, v = numpy.array(a, ndmin=1), numpy.array(v, ndmin=1)
+  if len(v) > len(a):
+    a, v = v, a
```
Need to use whitespace between code blocks for readability.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+
+def _rickerWavelet(points, a):
```
Here and elsewhere, the short arg and variable names like `a` are not conventional in python. Please use more descriptive names for code readability. Pylint doesn't like them either.

vitaly-krugl commented:
Function documentation is great overall, but need to follow conventions, which differ for nupic (Doxygen) and numenta-apps (Sphinx) code.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+               valueIndex=1):
+    """
+    :param fileName: str, path to input csv file
+    :param rowOffset: int, index of first data row in csv
```
add "zero-based"

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+class paramFinder(object):
+  def __init__(self,
+               fileName,
+               rowOffset=1,
```
Let's not have default values for any of these parameters. There are no really good, intuitive defaults for these, so it's better to force users to specify them explicitly. There are several reasons for this.

1. CSV files contains metric data used elsewhere in nupic (e.g., https://github.com/numenta/nupic/blob/master/src/nupic/data/file_record_stream.py) actually expect 3 header rows (not one).
2. There is no convention that timestamp should be in column 0 and value in column 1.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+
+class paramFinder(object):
```
By convention, class names should be mixed case, starting with upper case; so, this should be `ParamFinder`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    :param valueIndex: int, column index of the value
+    """
+
+    self._thresh = 0.2
```
Use a more descriptive name and/or add a comment that describes what this threshold applies to.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+                                           self._valueIndex)
+    numDataPts = len(value)
+
+    (self.medianSamplingInterval,
```
Here and elsewhere in this PR: define all instance member variables in `__init__` first. If there is no meaningful value for them in `__init__`, then simply assign `None`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+      self.medianSamplingInterval, value)
+
+    self.suggestedSamplingInterval = self.determineAggregationWindow(
+      self._timeScale,
```
When more than 1-2 args, pass then as keyword args. Error-prone and often not readable otherwise.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+  @staticmethod
+  def readCSVFiles(fileName,
```
readCSVFiles ==> readCSVFile (only reads one file)

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    """
+    Run paramFinder on input csv file
+
+    :return: JSON object with the following properties
```
return doc is incomplete, missing modelInfo

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    self._valueIndex = valueIndex
+
+
+  def run(self):
```
Based on what this does, I think it might be better off as a standalone function, and no class is needed. If I understand it correctly, it would be simpler for the user to just call a single function instead of constructing an instance of `ParamFinder` and calling `run`; e.g., `def findModelParams(fileName, rowOffset, timestampIndex, valueIndex)`). You can still subdivide the work among private functions and pass the necessary parameters by value.

This implementation is probably better served by functional programming than by OO programming; for example, `getAggregationFunction ()` depends on `self.medianAbsoluteDevSamplingInterval` and other instance members being set up by something else, so it's confusing to follow the code, and not clear in what order functions can legally be called.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    return useTimeOfDay, useDayOfWeek
+
+
+  def getAggregationFunction(self):
```
This and other private methods in this class need to use the naming convention for private functions, using a leading underscore. It would be inappropriate to call this as a public API, because it depends on `run()` to set up `self.medianAbsoluteDevSamplingInterval` and other instance variables.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    numDataPts = len(value)
+
+    (self.medianSamplingInterval,
+     self.medianAbsoluteDevSamplingInterval) = self.getMedianSamplingInterval(
```
It seems that `self.medianAbsoluteDevSamplingInterval` and other instance members here should use the naming convention for private members (leading underscrore)

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    :param timestampIndex: int, column index of the timeStamp
+    :param valueIndex: int, column index of the value
+    :return: timeStamps: numpy array of time stamps
+             values: numpy array of data values
```
The return value needs to make it clear that a two-tuple is returned; for example: 

A two-tuple (timestamps, values), where timestamps is a numpy array of `datetime64` time stamps, and values is a numpy array of `float32` data values as

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+          break
+
+      timeStamps = numpy.array(timeStamps, dtype='datetime64[s]')
+      values = numpy.array(values, dtype='float32')
```
Is float32 sufficient? This needs to be consistent with the maximum value size supported by nupic models for scalar values in metric data.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+      timeStamps = []
+      values = []
+
+      maxRowNumber = 20000
```
`maxRowNumber` should instead be defined as a class-level or module-level constant. Among other things, this would make it possible to patch in unit tests.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+class paramFinder(object):
+  def __init__(self,
+               fileName,
```
Expecting a CSV file in an API like this limits its flexibility. This means that anyone who needs to use this functionality needs to create a CSV file first in order to use this functionality. If you anticipate that this API would not need to scan an entire file (regardless of size), then it's more flexible to expose a constant that lets the users know the desired numbers of rows, and then accept the info as a sequence of (timestamp, value) two-tuples, and do away with all the other args here.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+          "func": A string representation of the aggregation function (string)
+            the following values are supported: "sum", "mean"
+    """
+    (timeStamp, value) = self.readCSVFiles(self._fileName,
```
These vars should be plural `timeStamps` and `values`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+  @staticmethod
+  def resampleData(timeStamp, value, newSamplingInterval):
```
Based on use, `totalDuration = (timeStamp[-1] - timeStamp[0])`, it appears that `timeStamp` is actually an array of timestamps, so need a more appropriate name for the `timeStamp` arg and also update param documentation.

Same for `value` ==> `values` and update param documentation.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+  @staticmethod
+  def calculateContinuousWaveletTransform(samplingInterval, value):
```
`value` ==> `values` and update docstring

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+
+  @staticmethod
+  def getMedianSamplingInterval(timeStamp):
```
`timeStamp` ==> `timeStamps` and update docstring

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+    medianSamplingInterval = numpy.median(samplingIntervals)
+    medianAbsoluteDev = numpy.median(numpy.abs(
+      samplingIntervals - medianSamplingInterval))
```
pylint probably won't like this type of hanging indentation. The following would be better:

```
medianAbsoluteDev = numpy.median(
  numpy.abs(samplingIntervals - medianSamplingInterval))
```
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    calculate median and median absolute deviation of sampling interval
+
+    :param timeStamp: a numpy array of sampling times
+    :return: medianSamplingInterval, numpy timedelta64 format in unit of seconds
```
Start by saying that it's a two-tuple

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    cwtmatr = _cwt(value, _rickerWavelet, widths)
+    # clip wavelet coefficients to minimize boundary effect
+    maxTimeScale = int(widths[-1])
+    cwtmatr = cwtmatr[:, 4 * maxTimeScale:-4 * maxTimeScale]
```
What does the comma mean in `cwtmatr[:, `?

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    timeScale = widths * samplingInterval * 4
+
+    # continuous wavelet transformation with ricker wavelet
+    cwtmatr = _cwt(value, _rickerWavelet, widths)
```
"matr" may be a non-standard abbreviation for matrix. It's not desirable to use non-standard abbreviations. `cwtMatrix` would be better.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+    :param samplingInterval: sampling interval of the time series
+    :param value: value of the time series
+    :return cwtVar: numpy array of flaots, variance of the wavelet coeff
+    :return timeScale: numpy array of floats, corresponding time scales
```
This is not a standard way to document a two-tuple return value. There is a single object being returned that is a two-tuple.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+      aggregationTimeScale = samplingInterval * 4
+
+    # make sure there is at least 1000 records after aggregation
+    if numDataPts < 1000:
```
This `1000` should probably be a class-level or module-level constant.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+
+    :param cwtVar: numpy array, wavelet coefficients variance over time
+    :param timeScale: numpy array, corresponding time scales for wavelet coeffs
+    """
```
Document return value

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -0,0 +1,163 @@
+
+  parser.add_argument("--rowOffset",
+                      type=int,
+                      default=1,
```
As explained in the review of `param_finder.py`, there are not really any good, intuitive defaults for rowOffset, timestampIndex, and valueIndex. It's better to make them required and not specify defaults.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,536 @@
+             values: numpy array of data values
+    """
+
+    with open(fileName, 'r') as csvFile:
```
The opening of the file and creation of the csv reader need a few more nuances. For creation of csv reader, see https://github.com/numenta/numenta-apps/blob/64fcd471fb79b97909537179528734256fcb7333/unicorn/py/unicorn_backend/model_runner_2.py#L250-L257.

For opening of the file, you need to specify "universal end-of-line" to work with Mac OS X, Windows (and linux) CSV files. Use "rU" as the open mode (instead of 'r').

vitaly-krugl commented:
Hi @ywcui1990, I am done with this round of code review. I focused primarily on code structure and coding/naming conventions. I am not familiar with the algorithms in question, so I didn't review them for correctness.

The main structural feedback for `param_finder.py` is:
1. The public interface to param finder should be a function rather than class. See https://github.com/numenta/numenta-apps/pull/571#discussion_r51220049 for explanation.
2. Accepting a CSV file as input is not desirable; see https://github.com/numenta/numenta-apps/pull/571/files#r51221239.

ywcui1990 commented:
@vitaly-krugl  Thanks very much for the detailed code review. I will take a thorough look at your comments tomorrow

ywcui1990 commented:
@marionleborgne @vitaly-krugl 

I have incorporated Vitaly's comments and added some unit tests. 

Summary of major changes
* Make param_finder a set of functions, instead of a class
* param_finder now takes timeStamps and values as input, instead of the csv file. The readCSVFile function has been moved to param_finder_runner
* No default values are given for any of the input parameters for param_finder_runner, the user has to provide all of them
* Add "timestampFieldName" and "valueFieldName" to output of param_finder, which are now hard coded as "c0" and "c1" respectively

Example usage:
python param_finder_runner.py --csv example_data/art_daily_no_noise.csv --rowOffset 1 --timestampIndex 0 --valueIndex 1 --datetimeFormat "%Y-%m-%d %H:%M:%S"

python param_finder_runner.py --csv [PATH_TO_NUPIC]/examples/opf/clients/hotgym/anomaly/one_gym/rec-center-hourly.csv --rowOffset 3 --timestampIndex 0 --valueIndex 1 --datetimeFormat "%m/%d/%y %H:%M"

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,529 @@
+  """
+  find parameters for a given time series dataset with heuristics
+
+  :param timeStamps: array_like, each entry is a timeStamp in datetime64 type
```
`datetime64`: I think that this is a numpy-specific data type. I don't think that we should expose our internal implementation details to users of the API. I think it would be more standard to take the builtin datetime.datetime type. The function would need to convert to whatever implementation-dependent type it needs.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,529 @@
+_CORRELATION_MODE_FULL = 2
+
+_AGGREGATION_WINDOW_THRESH = 0.2
+_MAX_ROW_NUMBER = 20000
```
If this is a suggested number of rows that users should pass, then this constant should be made public (remove leading underscore) and `find_parameters` docstring should refer to the constant in the explanation of using the API.

lscheinkman commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,529 @@
+                          aggFunc),
+    "modelInfo": getModelParams(useTimeOfDay, useDayOfWeek, values),
+    "timestampFieldName": "c0",
+    "valueFieldName": "c1"
```
@ywcui1990 `timestampFieldName` and `valueFieldName` should be properties of `modelInfo`, see https://github.com/numenta/numenta-apps/blob/master/unicorn/py/unicorn_backend/model_opt_schema.json

vitaly-krugl commented:
@ywcui1990, I started the review last night, but have not finished it. Please don't push changes to this PR until I let you know that I am done with the review. Thx.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+  :param vector2: (M,) array_like
+      Second one-dimensional input array.
+  :param mode: int that indicates mode of convolution, which takes the value of
+      2:
```
Use `_CORRELATION_MODE_FULL`, etc. instead of 2, etc. in documentation.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+MAX_ROW_NUMBER = 20000
+
+
+def _convolve(vector1, vector2, mode=2):
```
Use `_CORRELATION_MODE_FULL` instead of `2` in mode default value

vitaly-krugl replied:
Is `_CORRELATION_MODE_FULL` an intuitive default value for `_convolve`? In our own code, we are calling it with `_CORRELATION_MODE_SAME`. Since it's a private method, and we're actually calling it with an explicit `mode` that differs from default, let's get rid of the default value altogether.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+  Parameters
+  ----------
+  :param points: int
```
Using Sphinx param documentation, this would be `:param int points:` for types like int or datetime.datetime. If the type is more complex, such that it requires multiple words with spaces, then you would use the following two-line syntax:

```
:param dates: metric dates
:type dates: sequence of datetime.datetime objects
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+  Parameters
+  ----------
+  :param data : (N,) ndarray
```
In Sphinx notation, there is no space between parameter name and `:`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+      Array of length `points` in shape of ricker curve.
+
+  """
+  A = 2 / (numpy.sqrt(3 * waveletWidth) * (numpy.pi ** 0.25))
```
Per convention, variable names must start with lower-case letter and should be meaningful names. Single-letter names are not allowed.

Please run all code in this script through pylint using nupic/pylintrc configuration and see what else it finds and fix it. I am suggesting nupic/pylintrc for this script, because my understanding is that it will be migrated to nupic. If that's not the case, then use numenta-apps/pylintrc instead.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+  widths : (M,) sequence
+      Widths to use for transform.
+
+  :return Returns
```
`:return Returns` is neither Sphinx nor Doxygen

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+  """
+  find parameters for a given time series dataset with heuristics
+
+  :param timeStamps: array_like, each entry is a timeStamp in datetime type
```
What is `array_like`? Can they pass a python list or tuple? In that case, you probably want to say:

```
:param timeStamps: metric data timestamps
:type timeStamps: sequence of datetime.datetime timestamps
``

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+
+
+def find_parameters(timeStamps, values):
```
I think it would be cleaner and less error-prone for the API to accept a single sequence of timestamp/value two-tuples. Then there wouldn't be concerns that one might be longer than the other, etc. It would also be easier for users to use Python List Comprehension to build the parameter instead of having to use a for-loop.

vitaly-krugl replied:
Do we need to be concerned about noisy data, where some values might be missing (e.g., None) or `NaN`?

CC @scottpurdy @marionleborgne

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+    raise ValueError('timeStamps and Values must have the same length')
+
+  try:
+    assert(type(timeStamps[0]) is datetime.datetime)
```
Instead of assert, you probably want to do something similar to `ValueError` above, except use the `TypeError` exception.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+      timeStamps = timeStamps[:MAX_ROW_NUMBER]
+
+    timeStamps = numpy.array(timeStamps, dtype='datetime64[s]')
+  except:
```
This is too broad of a catch. Pylint wouldn't like it either. It will masquerade other problems/bugs as `TypeError`, making it confusing to debug. You want to catch more specific exceptions classes.

Also, you have at least two statements in the block that would cause the `TypeError` exception to be raised below. This would make it difficult for the user to debug whether the problem was in the first one `assert(type(timeStamps[0]) is datetime.datetime)` or the second one `numpy.array(timeStamps, dtype='datetime64[s]')`.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+      values = values[:MAX_ROW_NUMBER]
+
+    values = numpy.array(values).astype('float64')
+  except:
```
See prior feedback about "too broad of a catch".

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+    raise TypeError('timeStamps must be an array of datetime64')
+
+  try:
+    if len(values) > MAX_ROW_NUMBER:
```
This doesn't need to be in the try-except block.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+
+
+def getAggInfo(medianSamplingInterval, suggestedSamplingInterval, aggFunc):
```
Is this and the other functions in this script, except `find_parameters`, supposed to be public? If not, need to add leading underscore to mark them as private.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+
+
+def getModelParams(useTimeOfDay, useDayOfWeek, value):
```
Rename `value` as `values`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+  Return a JSON object describing the model configuration
+  :param useTimeOfDay: bool, whether to use timeOfDay encoder
+  :param useDayOfWeek: bool, whether to use dayOfWeej encoder
+  :param value: numpy array of metric data, used to compute min/max values
```
In nupic's developer's mind, metric data usually includes a timestamp and value. So the use of "metric data" in this param description is confusing. Please use a more specific/accurate description.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+def getModelParams(useTimeOfDay, useDayOfWeek, value):
+  """
+  Return a JSON object describing the model configuration
+  :param useTimeOfDay: bool, whether to use timeOfDay encoder
```
`:param bool useTimeOfDay: whether to use timeOfDay encoder`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+  :param timeStamps: array_like, each entry is a timeStamp in datetime type
+  :param values: array_like, each entry is a float number
+
+  :return: JSON object with the following properties:
```
In Sphinx, this would be `:returns:`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+    "modelInfo": JSON object describing the model configuration that
+    contains the following properties:
+      “modelConfig”: OPF Model Configuration parameters (JSON object) for
```
Want to stick with ASCII quotes to avoid compatibility issues. Search your code for other uses of these quotes.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+  if useTimeOfDay:
+    modelParams['modelConfig']['modelParams']['sensorParams']['encoders'] \
+      ['c0_timeOfDay'] = dict(fieldname='c0',
```
The logic here is very tightly coupled with `getScalarMetricWithTimeOfDayAnomalyParams`, making this implementation brittle. This implementation could easily break in subtle, unpredictable ways if object returned by `getScalarMetricWithTimeOfDayAnomalyParams` changes in certain ways.

For example, if `getScalarMetricWithTimeOfDayAnomalyParams` renames `c0_timeOfDay`, the result would not be meaningful.

Need logic (possibly in tests) to detect this.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+    aggregationTimeScale = samplingInterval * 4
+
+  # make sure there is at least 1000 records after aggregation
+  if numDataPts < 1000:
```
Use a module-level named constant instead of `1000` here. This would enable unit tests to use the constant to test each branch.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+        useTimeOfDay = True
+
+      if (timeScale[leftLocalMin] < weekPeriod < timeScale[
+        rightLocalMin]
```
The indentation here doesn't look right. Try the following

```
if (timeScale[leftLocalMin] < weekPeriod < 
    timeScale[rightLocalMin] and
    cwtVarAtWeekPeriod > localMaxValue * 0.5):
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+      if (timeScale[leftLocalMin] < weekPeriod < timeScale[
+        rightLocalMin]
+          and cwtVarAtWeekPeriod > localMaxValue * 0.5):
```
The indentation here doesn't look right. Try the following

```
if (timeScale[leftLocalMin] < weekPeriod < 
    timeScale[rightLocalMin] and
    cwtVarAtWeekPeriod > localMaxValue * 0.5):
```

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -0,0 +1,242 @@
+
+      # use local timezone to be consistent with the default assumption
+      # in numpy datetime64 object
+      timeStamp = timeStamp.replace(tzinfo=tz.tzlocal())
```
This may result in invalid date that would cause an exception at some point in execution. For example, if the original timestamp falls in the non-existent one-hour gap that happens during spring-time daylight savings time adjustment, then the resulting timestamp becomes invalid.

Perhaps a UTC time zone would be a better choice.

Also, add a data sample to test for this.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -0,0 +1,242 @@
+  """
+
+  with open(fileName, 'rU') as csvFile:
+    csv.field_size_limit(2 ** 27)
```
Also need to tell the system to use "\n" as line separators for Windows compatibility. See https://github.com/numenta/numenta-apps/blob/64fcd471fb79b97909537179528734256fcb7333/unicorn/py/unicorn_backend/model_runner_2.py#L254-L255

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -0,0 +1,242 @@
+
+g_log = logging.getLogger(__name__)
+
+_MAX_ROW_NUMBER = 20000
```
Don't define your own max row number. Use the public one from param_finder.py instead.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+_CORRELATION_MODE_FULL = 2
+
+_AGGREGATION_WINDOW_THRESH = 0.2
+MAX_ROW_NUMBER = 20000
```
`MAX_NUM_ROWS` would be a better name. Also, add a comment that explains what it's for

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder_runner.py:
```diff
@@ -0,0 +1,242 @@
+
+  parser = SilentArgumentParser(description=("Start Unicorn ParamFinderRunner"))
+
+  parser.add_argument("--csv",
```
These options don't match the interface specification for Param Finder. See "ParamFinder Command-line Args" in https://docs.google.com/document/d/1XZYGFCIoK7NarAl-DzCqluaTHac1PYFZQaEiu0r2rhs/edit

vitaly-krugl commented on unicorn/py/unicorn_backend/tests/unit/param_finder_runner_test.py:
```diff
@@ -0,0 +1,126 @@
+
+
+    _assertArgumentPatternFails()
+    _assertArgumentPatternFails(["--csv="])
```
See comment about invalid options https://github.com/numenta/numenta-apps/pull/571#discussion_r51661994

vitaly-krugl commented on unicorn/py/unicorn_backend/tests/unit/param_finder_runner_test.py:
```diff
@@ -0,0 +1,126 @@
+    """
+    Verify CSV file can be correctly read via  _readCSVFile function
+    """
+    tmpfilepath = os.path.join(tempfile.gettempdir(), "tmp-testfile.csv")
```
Use mkdtemp instead (safe) with immediate addCleanup like this https://github.com/numenta/numenta-apps/blob/72ea1f688fd5309dc00445fa2bcaf06f161571b4/htm.it/tests/py/integration/app/runtime/metric_collector_int_test.py#L121-L123

and get rid of os.remove at the bottom

vitaly-krugl commented on unicorn/py/unicorn_backend/tests/unit/param_finder_runner_test.py:
```diff
@@ -0,0 +1,126 @@
+                           numpy.datetime64('2014-04-01 00:00:00'))
+    os.remove(tmpfilepath)
+
+
```
Need end-to-end test that calls param finder runner's `main` with the appropriate `sys.argv` and then validates the value written to stdout.

For this, 

1. Create a temp csv file
2. patch `sys.argv` with appropriate args
3. patch sys.stdout with `StringIO.StringIO` object
4. Call `main`
5. After `main` returns, verify that the `StringIO.StringIO` object has the expected values
6. Verify that the `StringIO.StringIO` object's contents match the schema for the return value (you will need to define a JSON schema for param finder's result similar to the ones that I defined for ModelRunner's command-line option values)

vitaly-krugl commented on unicorn/py/unicorn_backend/tests/unit/param_finder_test.py:
```diff
@@ -0,0 +1,175 @@
+    self.assertLessEqual(aggregationTimeScale, maxSamplingInterval)
+    self.assertGreater(aggregationTimeScale, samplingInterval.astype('float64'))
+
+
```
Missing test for `find_parameters`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+    cwtVar=_cwtVar,
+    thresh=_AGGREGATION_WINDOW_THRESH,
+    samplingInterval=medianSamplingInterval,
+    numDataPts=numDataPts)
```
Need to test both code paths: below numDataPts threshold and above. See https://github.com/numenta/numenta-apps/pull/571#discussion_r51649813

vitaly-krugl commented:
@ywcui1990, I am done with this round of code review. Please seem my comments for issues that need to be addressed.

I usually did not repeat comments for an issue that occurs in multiple places for things like coding/naming conventions. For example, if you see feedback about format of parameter or return value documentation, self-review and correct all parameter and/or return value documentation in all scripts in this PR that are not consistent with our convention.

Also, run pylint on all scripts, including tests, and fix what it finds. The pylint configuration file for nupic is is nupic/pylintrc and the one for numenta-apps is in numenta-apps/pylintrc.

CC @marionleborgne

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -0,0 +1,537 @@
+
+def find_parameters(timeStamps, values):
+  """
+  find parameters for a given time series dataset with heuristics
```
Describe in docstring the expectation of `MAX_ROW_NUMBER` rows.

ywcui1990 commented:
@marionleborgne @lscheinkman 
Here is a version of the code that follows the technical requirements with the correct input/output. I am still addressing Vitaly's comments on internal implementations and coding styles. But this version should unblock you for now.

vitaly-krugl commented on unicorn/py/unicorn_backend/input_opt_schema_param_finder.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "description": "--input command-line option object passed to model runner",
```
Need to update descriptions that refer to model-runner-specifics.

vitaly-krugl commented on unicorn/py/unicorn_backend/input_opt_schema_param_finder.json:
```diff
@@ -0,0 +1,37 @@
+
+  "properties": {
+    "csv": {
+      "description": "Absolute path to input CSV file; if omitted, will read CSV from stdin.",
```
"if omitted, will read CSV from stdin" doesn't belong here, since "csv" path in not optional in param finder runner

marionleborgne commented:
FYI, currently testing the script. Reported errors to @ywcui1990

marionleborgne commented:
:+1:  it works. Thanks @ywcui1990. You can merge this one and open another PR to address @vitaly-krugl 's comments.

ywcui1990 commented:
@marionleborgne I don't have write access to this repository. Feel free to merge it on my behalf. I will address the rest of @vitaly-krugl 's comments in another PR.

marionleborgne commented:
:+1: great. Thanks @ywcui1990
